### PR TITLE
glibc (new formula)

### PR DESCRIPTION
**Pull request by iMichka**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
`glibc` needs to stay at version 2.23. It is only needed for users on linux that work on systems that have too old glibc versions, so we just provide this as a nicety for them. The version is aligned with the glibc version on our CI, so that our bottles stay compatible.

**Comment by iMichka** [comment]
> ==> SKIPPED glibc
gawk 3.1.2 (or later) is required to build glibc.
Install gawk with your host package manager if you have sudo access:
  sudo apt-get install gawk
  sudo yum install gawk

I need to add gawk to our CI docker image

File: Formula/glibc.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    (prefix/"etc").install_symlink sys_localtime if sys_localtime.exist? && !brew_localtime.exist?
+
+    # Set zoneinfo correctly using the system installed zoneinfo
+    sys_zoneinfo = Pathname.new "/usr/share/zoneinfo"
+    brew_zoneinfo = Pathname.new share/"zoneinfo"
+    share.install_symlink sys_zoneinfo if sys_zoneinfo.exist? && !brew_zoneinfo.exist?
+  end
+
+  test do
+    system "#{lib}/ld-#{version}.so 2>&1 |grep Usage"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:171]: `assert_match` rather than `grep` here.

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by iMichka** [comment]
Needs https://github.com/Homebrew/homebrew-linux-dev/pull/196

**Comment by iMichka** [comment]
Actually https://github.com/Homebrew/brew/pull/10732 is needed.

**Comment by iMichka** [comment]
```
2021-03-02T06:53:34.4815962Z make[2]: *** No rule to make target '/tmp/glibc-20210302-31647-r36ljy/glibc-2.23/build/mathvec/libmvec.so.1', needed by '/home/linuxbrew/.linuxbrew/Cellar/glibc/2.23/lib/libm.so'.  Stop.
2021-03-02T06:53:34.4817076Z make[2]: *** Waiting for unfinished jobs....
2021-03-02T06:53:34.4818053Z make[2]: Leaving directory '/tmp/glibc-20210302-31647-r36ljy/glibc-2.23/math'
2021-03-02T06:53:34.4819410Z Makefile:214: recipe for target 'math/subdir_install' failed
2021-03-02T06:53:34.4820045Z make[1]: *** [math/subdir_install] Error 2
2021-03-02T06:53:34.4820940Z make[1]: Leaving directory '/tmp/glibc-20210302-31647-r36ljy/glibc-2.23'
2021-03-02T06:53:34.4821876Z Makefile:12: recipe for target 'install' failed
2021-03-02T06:53:34.4822419Z make: *** [install] Error 2
```

File: Formula/glibc.rb [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      ]
+      system "../configure", *args
+      system "make", "all"
+      system "make", "install"
+      prefix.install_symlink "lib" => "lib64"
+    end
+  end
+
+  def post_install
+    # Fix permissions
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:138]: `skip_clean` might do this for you? Would be good to get this in `install` instead if possible.

File: Formula/glibc.rb [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    chmod 0755, [lib/"ld-#{version}.so", lib/"libc-#{version}.so"]
+
+    # Install ld.so symlink.
+    ln_sf lib/"ld-linux-x86-64.so.2", HOMEBREW_PREFIX/"lib/ld.so"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:142]: Same here: can this be in `install`?

File: Formula/glibc.rb [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    # Compile locale definition files
+    mkdir_p lib/"locale"
+    locales = ENV.map { |k, v| v if k[/^LANG$|^LC_/] && v != "C" && !v.start_with?("C.") }.compact
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:146]: Can this be a multiline block and have a comment explaining what's going on here.

File: Formula/glibc.rb [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        charmap = "UTF-8" if charmap == "utf8"
+        system bin/"localedef", "-i", lang, "-f", charmap, locale
+      else
+        system bin/"localedef", "-i", lang, locale
+      end
+    end
+
+    # Set the local time zone
+    sys_localtime = Pathname.new "/etc/localtime"
+    brew_localtime = Pathname.new prefix/"etc/localtime"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:162]:
```suggestion
    brew_localtime = prefix/"etc/localtime"
```

File: Formula/glibc.rb [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    (prefix/"etc").install_symlink sys_localtime if sys_localtime.exist? && !brew_localtime.exist?
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:163]: What if this symlink exists but it's pointing to the wrong place?

File: Formula/glibc.rb [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+    # Set zoneinfo correctly using the system installed zoneinfo
+    sys_zoneinfo = Pathname.new "/usr/share/zoneinfo"
+    brew_zoneinfo = Pathname.new share/"zoneinfo"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:167]:
```suggestion
    brew_zoneinfo = share/"zoneinfo"
```

File: Formula/glibc.rb (hunk 2 above)
Comment (iMichka) [reply Formula/glibc.rb]: I removed that chmod call, it looks like it's not needed.

File: Formula/glibc.rb (hunk 3 above)
Comment (iMichka) [reply Formula/glibc.rb:137]: I moved this to the install step.

File: Formula/glibc.rb (hunk 6 above)
Comment (iMichka) [reply Formula/glibc.rb]: I guess that if `/etc/localtime` is missing or points to the wrong place, you might have bigger issues with your system.
https://www.freedesktop.org/software/systemd/man/localtime.html

It looks like a quite common file on unix. On ubuntu 16.04:
> lrwxrwxrwx 1 root root 27 May 25  2020 /etc/localtime -> /usr/share/zoneinfo/Etc/UTC

File: Formula/glibc.rb [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    end
+
+    # Install ld.so symlink.
+    ln_sf lib/"ld-linux-x86-64.so.2", HOMEBREW_PREFIX/"lib/ld.so"
+  end
+
+  def post_install
+    # Compile locale definition files
+    mkdir_p lib/"locale"
+    # Get all extra installed locales from the system, except C locales
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:143]:
```suggestion
    mkdir_p lib/"locale"
    
    # Get all extra installed locales from the system, except C locales
```

File: Formula/glibc.rb [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    locales = ENV.map do |k, v|
+      v if k[/^LANG$|^LC_/] && v != "C" && !v.start_with?("C.")
+    end
+    locales = locales.compact
+    # en_US.UTF-8 is required by gawk make check
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:148]:
```suggestion
    end.compact

    # en_US.UTF-8 is required by gawk make check
```

File: Formula/glibc.rb [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+      if charmap.present?
+        charmap = "UTF-8" if charmap == "utf8"
+        system bin/"localedef", "-i", lang, "-f", charmap, locale
+      else
+        system bin/"localedef", "-i", lang, locale
+      end
+    end
+
+    # Set the local time zone
+    sys_localtime = "/etc/localtime"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:162]:
```suggestion
    sys_localtime = Pathname("/etc/localtime")
```

File: Formula/glibc.rb [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    brew_localtime = prefix/"etc/localtime"
+    (prefix/"etc").install_symlink sys_localtime if Dir.exist?(sys_localtime) && !Dir.exist?(brew_localtime)
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:164]:
```suggestion
    (prefix/"etc").install_symlink sys_localtime if sys_localtime.exist? && !brew_localtime.exist?
```

File: Formula/glibc.rb [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+    # Set zoneinfo correctly using the system installed zoneinfo
+    sys_zoneinfo = "/usr/share/zoneinfo"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:167]:
```suggestion
    sys_zoneinfo = Pathname("/usr/share/zoneinfo")
```

File: Formula/glibc.rb [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    brew_zoneinfo = share/"zoneinfo"
+    share.install_symlink sys_zoneinfo if Dir.exist?(sys_zoneinfo) && !Dir.exist?(brew_zoneinfo)
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:169]:
```suggestion
    share.install_symlink sys_zoneinfo if sys_zoneinfo.exist? && !brew_zoneinfo.exist?
```

File: Formula/glibc.rb [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+  end
+
+  test do
+    assert_match "Usage", shell_output("#{lib}/ld-#{version}.so 2>&1", 127)
+    system "#{lib}/libc-#{version}.so", "--version"
+    system "#{bin}/locale", "--version"
```
Comment (MikeMcQuaid) [inline Formula/glibc.rb:175]:
```suggestion
    safe_system "#{lib}/libc-#{version}.so", "--version"
    safe_system "#{bin}/locale", "--version"
```

or 

```suggestion
    assert_equal true, system "#{lib}/libc-#{version}.so", "--version"
    assert_equal true, system "#{bin}/locale", "--version"
```

File: Formula/glibc.rb (hunk 10 above)
Comment (iMichka) [reply Formula/glibc.rb:162]: I think you made me remove the Pathnames in the previous comments :D

File: Formula/glibc.rb (hunk 9 above)
Comment (iMichka) [reply Formula/glibc.rb:148]: Cool I did not know I could do that.

File: Formula/glibc.rb (hunk 10 above)
Comment (iMichka) [reply Formula/glibc.rb]: Ha gotcha it was only for the second variable.

Comment (MikeMcQuaid) [reply Formula/glibc.rb]: Yup as it's implicitly a Pathname 😅

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**